File: kotlin-workshop/src/main/kotlin/nl/sourcelabs/solutions/E07_ObjectVsClassFromJava.java
package nl.sourcelabs.solutions;

public class E07_ObjectVsClassFromJava {
    public static void main(String[] args) {
        /**
         * Invoke HelloworldO sayHello
         * Invoke HelloworldO sayHello("yourname")
         */
        HelloworldOE07.INSTANCE.sayHello(); // Generated because of the @JvmOverloads
        HelloworldOE07.INSTANCE.sayHello("Stephan");

        // Bonus: static by specifying @JvmStatic on a object function

        HelloworldOE07.sayHello(); // Generated because of the @JvmOverloads
        HelloworldOE07.sayHello("Stephan");

        /**
         * Invoke HelloworldC sayHello
         * Invoke HelloworldC sayHello("yourname")
         */
        new HelloworldCE07().sayHello();
        new HelloworldCE07().sayHello("Stephan");
    }
}

File: kotlin-workshop/src/main/kotlin/nl/sourcelabs/solutions/E07_ObjectVsClassFromJava.kt
package nl.sourcelabs.solutions

/**
 * Define an object called HelloworldO.
 *
 * Write a member function sayHello(name) that prints a hello name
 *
 * Invoke the function static from Java
 */
object HelloworldOE07 {
    @JvmStatic // allows for static invocation from Java
    @JvmOverloads // generates function for Java with the default value provided
    fun sayHello(name: String = "world") {
        println("Hello: $name from $this")
    }
}

/**
 * Define a class called HelloworldC.
 *
 * Write a member function sayHello(name) that prints a hello name
 */
class HelloworldCE07 {
    @JvmOverloads // generates function for Java with the default value provided
    fun sayHello(name: String = "world") {
        println("Hello $name for $this")
    }
}
